File: Lua/Vector2d.lua
local Vector2d = {}

function Vector2d.pointToPointVector(from, to)

	local vectorOfMovement = {x, y}
	vectorOfMovement.x  = to.x - from.x
	vectorOfMovement.y = to.y - from.y

	return vectorOfMovement
end	

function Vector2d.vectorFromRotationInRadians(rotationInRadians, velocity)
	
	local vectorOfMovement = {x, y}
	
	if(velocity ~= nil) then
		vectorOfMovement.x = math.sin(rotationInRadians) * velocity
		vectorOfMovement.y = -math.cos(rotationInRadians) * velocity
	else 
		vectorOfMovement.x = math.sin(rotationInRadians)
		vectorOfMovement.y = -math.cos(rotationInRadians)
	end
	return vectorOfMovement
end

function Vector2d.vectorFromRotationInDegrees(rotationInDegrees, velocity)
	
	local vectorOfMovement = {x, y}
	local rotationInRadians = rotationInDegrees * math.pi / 180
	
	if(velocity ~= nil) then
		vectorOfMovement.x = math.sin(rotationInRadians) * velocity
		vectorOfMovement.y = -math.cos(rotationInRadians) * velocity
	else 
		vectorOfMovement.x = math.sin(rotationInRadians)
		vectorOfMovement.y = -math.cos(rotationInRadians)
	end

	return vectorOfMovement
end

function Vector2d.rotationInDegreesFromVector(vectorOfMovement)
	
	local angleInRadians = math.atan2(vectorOfMovement.y, vectorOfMovement.x)
	local angleInDegrees = (angleInRadians * 180) / math.pi

	return angleInDegrees
end

function Vector2d.rotationInRadiansFromVector(vectorOfMovement)
	
	local angleInRadians = math.atan2(vectorOfMovement.y, vectorOfMovement.x)

	return angleInRadians
end

function Vector2d.radiansToDegrees(radians)
	local degrees = ( radians * 180 ) / math.pi

	return degrees
end

function Vector2d.degreesToRadians(degrees)
	local radians = ( degrees * math.pi ) / 180

	return radians
end

return Vector2d

File: Java/ml/tycjan/vector2d/Vector2d.java
package ml.tycjan.vector2d;

public class Vector2d {
	
	private float x;
	private float y;
	
	public Vector2d(float x, float y){
		this.x = x;
		this.y = y;
	}
	
	public void setX(float x) {
		this.x = x;
	}

	public void setY(float y) {
		this.y = y;
	}

	public float getX() {
		return x;
	}

	public float getY() {
		return y;
	}
	
	
	public static float radiansToDegrees(float radians){
		
		float degrees = (float) (radians * Math.PI / 180f);
		return degrees;
	}
	
	public static float degreesToRadians(float degrees){
		
		float radians = (float) (degrees * 180f / Math.PI);
		return radians;
	}
	
	public static float rotationInRadiansFromVector(float toX, float toY, float fromX, float fromY){
		float angle = (float) (Math.atan2(toY - fromY, toX - fromX));
		float radians = angle;
		
		return radians;
	}
	
	public static float rotationInRadiansFromVector(Vector2d vector){
		float angle = (float) (Math.atan2(vector.y, vector.x));
		float radians = angle;
		
		return radians;
	}
	
	public static float rotationInDegreesFromVector(float toX, float toY, float fromX, float fromY){
		float angle = (float) (Math.atan2(toY - fromY, toX - fromX));
		float degrees = (float) (angle * 180f / Math.PI);
		
		return degrees;
	}
	
	public static float rotationInDegreesFromVector(Vector2d vector){
		float angle = (float) (Math.atan2(vector.y, vector.x));
		float degrees = (float) (angle * 180f / Math.PI);
		
		return degrees;
	}
	
	public static Vector2d pointToPointVector(float toX, float toY, float fromX, float fromY){
		
		Vector2d vector2d = new Vector2d(toX - fromX, toY - fromY);
		return vector2d;
	}
	
	public static Vector2d pointToPointVector(Vector2d to, Vector2d from){
		
		Vector2d vector2d = new Vector2d(to.x - from.x, to.y - from.y);
		return vector2d;
	}
	
	
	public static Vector2d vectorFromRotationInRadians(float radians){
		
		float vectorX = (float)(Math.sin(radians));
		float vectorY = (float)(-Math.cos(radians));
		
		Vector2d vector2d = new Vector2d(vectorX, vectorY);
		return vector2d;
	}
	
	public static Vector2d vectorFromRotationInDegrees(float degrees){
		
		float radians = Vector2d.degreesToRadians(degrees);
		
		float vectorX = (float)(Math.sin(radians));
		float vectorY = (float)(-Math.cos(radians));
		
		Vector2d vector2d = new Vector2d(vectorX, vectorY);
		return vector2d;
	}
	
}
